node/tessel-export.js

```diff
@@ -952,6 +957,251 @@ Tessel.LED.prototype.read = function(callback) {
+  restartWifi()
+    .then(() => {
+      this.busy = false;
+      this.connected = true;
+      this.emit('connect', this.settings);
+      callback();
+    })
+    .catch((error) => {
+      this.busy = false;
+      this.emit('error', error);
+      callback(error);
+    });
+};
+
+Tessel.Wifi.prototype.connection = function() {
+  if (this.isConnected) {
+    return this.settings;
+  } else {
+    return null;
+  }
+};
+
+Tessel.Wifi.prototype.connect = function(settings, callback) {
+  if (typeof settings !== 'object' || settings.ssid.length === 0) {
+    throw new Error('Wifi settings must be an object with at least a "ssid" property.');
+  }
+
+  if (typeof callback !== 'function') {
+    callback = function() {};
+  }
+
+  if (settings.password && !settings.security) {
+    settings.security = 'psk2';
+  }
+
+  if (!settings.password && (!settings.security || settings.security === 'none')) {
+    settings.security = 'none';
+  }
+
+  this.busy = true;
+  connectToNetwork(settings)
+    .then(turnOnWifi)
+    .then(commitWireless)
+    .then(restartWifi)
+    .then(getWifiInfo)
+    .then((network) => {
+      this.settings = Object.assign(network, settings);
+      this.busy = false;
+      this.connected = true;
+      this.emit('connect', this.settings);
+
+      callback(null, this.settings);
+    })
+    .catch((error) => {
+      this.busy = false;
+
+      this.emit('error', error);
+      callback(error);
+    });
+};
+
+function connectToNetwork(settings) {
+  var commands = `
+    uci batch <<EOF
+    set wireless.@wifi-iface[0].ssid=${settings.ssid}
+    set wireless.@wifi-iface[0].key=${settings.password}
+    set wireless.@wifi-iface[0].encryption=${settings.security}
+    EOF
+  `;
+
+  return new Promise((resolve) => {
+    childProcess.exec(commands, (error) => {
+      if (error) {
+        throw error;
+      }
+
+      resolve();
+    });
+  });
+}
+
+function turnOnWifi() {
+  return new Promise((resolve) => {
+    childProcess.exec('uci set wireless.@wifi-iface[0].disabled=0', (error) => {
+      if (error) {
+        throw error;
+      }
+      resolve();
+    });
+  });
+}
+
+function turnOffWifi() {
+  return new Promise((resolve) => {
+    childProcess.exec('uci set wireless.@wifi-iface[0].disabled=1', (error) => {
+      if (error) {
+        throw error;
+      }
+      resolve();
+    });
+  });
+}
+
+function commitWireless() {
+  return new Promise((resolve) => {
+    childProcess.exec('uci commit wireless', (error) => {
+      if (error) {
+        throw error;
+      }
+      resolve();
+    });
+  });
+}
+
+function restartWifi() {
+  return new Promise((resolve) => {
+    childProcess.exec('wifi', (error) => {
+      if (error) {
+        throw error;
+      }
+
+      resolve();
+    });
+  });
+}
+
+function getWifiInfo() {
+  return new Promise((resolve, reject) => {
+    childProcess.exec('ubus call iwinfo info {"device":"wlan0"}', (error, results) => {
```

Comment:
This command proved not very consistent when used with the CLI. See https://github.com/tessel/t2-cli/pull/517. We may want to use the less-friendly but consistent method I used there.